### PR TITLE
[SPARK-13027][STREAMING] Added batch time as a parameter to updateStateByKey

### PR DESCRIPTION
**Pull request by aramesh117**
Added RDD batch time as an input parameter to the update function in updateStateByKey.

### CODE REVIEW

**Comment by zsxwing** [comment]
ok to test

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/StateDStream.scala [hunk 1]
```diff
@@ -42,8 +42,8 @@ class StateDStream[K: ClassTag, V: ClassTag, S: ClassTag](
 
   override val mustCheckpoint = true
 
-  private [this] def computeUsingPreviousRDD (
-    parentRDD: RDD[(K, V)], prevStateRDD: RDD[(K, S)]) = {
+  private [this] def computeUsingPreviousRDD (batchTime: Time,
+                                              parentRDD: RDD[(K, V)], prevStateRDD: RDD[(K, S)]) = {
```
Comment (zsxwing) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/StateDStream.scala]: nit: code style. Should be 
```
private [this] def computeUsingPreviousRDD (
    batchTime: Time,
    parentRDD: RDD[(K, V)], 
    prevStateRDD: RDD[(K, S)]) = {
```

Comment (aramesh117) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/StateDStream.scala]: Will fix, thank you!

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/StateDStream.scala [hunk 2]
```diff
@@ -108,7 +109,8 @@ class StateDStream[K: ClassTag, V: ClassTag, S: ClassTag](
                 Some (sessionRDD)
               }
               case Some (initialStateRDD) => {
-                computeUsingPreviousRDD(parentRDD, initialStateRDD)
+                computeUsingPreviousRDD(validTime,
```
Comment (zsxwing) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/StateDStream.scala]: nit: could be fit into one line

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/PairDStreamFunctions.scala [hunk 3]
```diff
@@ -453,8 +453,14 @@ class PairDStreamFunctions[K, V](self: DStream[(K, V)])
       updateFunc: (Iterator[(K, Seq[V], Option[S])]) => Iterator[(K, S)],
       partitioner: Partitioner,
       rememberPartitioner: Boolean
-    ): DStream[(K, S)] = ssc.withScope {
-     new StateDStream(self, ssc.sc.clean(updateFunc), partitioner, rememberPartitioner, None)
+    ): DStream[(K, S)] = {
```
Comment (zsxwing) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/PairDStreamFunctions.scala]: nit: could you remove the outer `{}` like other methods?

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/PairDStreamFunctions.scala [hunk 4]
```diff
@@ -499,9 +505,35 @@ class PairDStreamFunctions[K, V](self: DStream[(K, V)])
       partitioner: Partitioner,
       rememberPartitioner: Boolean,
       initialRDD: RDD[(K, S)]
-    ): DStream[(K, S)] = ssc.withScope {
-     new StateDStream(self, ssc.sc.clean(updateFunc), partitioner,
-       rememberPartitioner, Some(initialRDD))
+    ): DStream[(K, S)] = {
```
Comment (zsxwing) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/PairDStreamFunctions.scala]: nit: could you remove the outer `{}` like other methods?

**Comment by zsxwing** [comment]
Could you add a method with `rememberPartitioner: Boolean`?

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/StateDStream.scala (hunk 2 above)
Comment (aramesh117) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/StateDStream.scala]: Done.

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/PairDStreamFunctions.scala (hunk 3 above)
Comment (aramesh117) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/PairDStreamFunctions.scala]: Done.

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/PairDStreamFunctions.scala (hunk 4 above)
Comment (aramesh117) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/PairDStreamFunctions.scala]: Done.

**Comment by aramesh117** [comment]
@zsxwing Sure.

File: streaming/src/test/scala/org/apache/spark/streaming/BasicOperationsSuite.scala [hunk 5]
```diff
@@ -467,6 +467,73 @@ class BasicOperationsSuite extends TestSuiteBase {
...
+                        state: Option[StreamingState]) => {
+        // Update only if we receive values for this key during the batch.
+        if (values.nonEmpty) {
+          Option(time.milliseconds + state.getOrElse(0L))
+        } else {
+          Option(state.getOrElse(0L))
+        }
+      }
+      val x = s.map(x => (x, 1))
+        x.updateStateByKey[StreamingState](updateFunc = updateFunc,
```
Comment (zsxwing) [inline streaming/src/test/scala/org/apache/spark/streaming/BasicOperationsSuite.scala]: nit: not need to create `x`. Just `s.map(x => (x, 1)).updateStateByKey`

Comment (aramesh117) [inline streaming/src/test/scala/org/apache/spark/streaming/BasicOperationsSuite.scala]: Sorry yes, I had created a variable to find its type while debugging an issue. I'll restore the original.

**Comment by aramesh117** [comment]
@zsxwing Any more issues that you can see?

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/PairDStreamFunctions.scala [hunk 6]
```diff
@@ -452,11 +452,15 @@ class PairDStreamFunctions[K, V](self: DStream[(K, V)])
...
-     new StateDStream(self, ssc.sc.clean(updateFunc), partitioner, rememberPartitioner, None)
+      rememberPartitioner: Boolean): DStream[(K, S)] = ssc.withScope {
+    val cleanedFunc = ssc.sc.clean(updateFunc)
+    val newUpdateFunc = (_: Time, it: Iterator[(K, Seq[V], Option[S])]) => {
+      cleanedFunc(it)
+    }
+    new StateDStream(self, newUpdateFunc, partitioner, rememberPartitioner, None)
   }
 
+
```
Comment (zsxwing) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/PairDStreamFunctions.scala]: nit: extra line

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/PairDStreamFunctions.scala [hunk 7]
```diff
@@ -498,10 +502,36 @@ class PairDStreamFunctions[K, V](self: DStream[(K, V)])
...
+    * the given function on the previous state of the key and the new values of the key.
+    * org.apache.spark.Partitioner is used to control the partitioning of each RDD.
+    * @param updateFunc State update function. If `this` function returns None, then
+    *                   corresponding state key-value pair will be eliminated.
+    * @param partitioner Partitioner for controlling the partitioning of each RDD in the new
+    *                    DStream.
+    * @tparam S State type
+    */
+  def updateStateByKey[S: ClassTag]
+  (updateFunc: (Time, K, Seq[V], Option[S]) => Option[S],
```
Comment (zsxwing) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/PairDStreamFunctions.scala]: nit: move `(` to the above line.

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/PairDStreamFunctions.scala [hunk 8]
```diff
@@ -498,10 +502,36 @@ class PairDStreamFunctions[K, V](self: DStream[(K, V)])
...
+   partitioner: Partitioner,
```
Comment (zsxwing) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/PairDStreamFunctions.scala]: Please use 4 spaces indention for the parameters

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/PairDStreamFunctions.scala [hunk 9]
```diff
@@ -498,10 +502,36 @@ class PairDStreamFunctions[K, V](self: DStream[(K, V)])
...
+   rememberPartitioner: Boolean,
+   initialRDD: Option[RDD[(K, S)]] = None): DStream[(K, S)] = ssc.withScope {
+    val cleanedFunc = ssc.sc.clean(updateFunc)
+    val newUpdateFunc = (time: Time, iterator: Iterator[(K, Seq[V], Option[S])]) => {
+      iterator.flatMap(t => cleanedFunc(time, t._1, t._2, t._3).map(s => (t._1, s)))
+    }
+    new StateDStream(self, newUpdateFunc, partitioner,
```
Comment (zsxwing) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/PairDStreamFunctions.scala]: nit: can be fit into one line

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/PairDStreamFunctions.scala [hunk 10]
```diff
@@ -498,10 +502,36 @@ class PairDStreamFunctions[K, V](self: DStream[(K, V)])
...
-    ): DStream[(K, S)] = ssc.withScope {
-     new StateDStream(self, ssc.sc.clean(updateFunc), partitioner,
-       rememberPartitioner, Some(initialRDD))
+      initialRDD: RDD[(K, S)]): DStream[(K, S)] = ssc.withScope {
+    val cleanedFunc = ssc.sc.clean(updateFunc)
+    val newUpdateFunc = (_: Time, it: Iterator[(K, Seq[V], Option[S])]) => {
+      cleanedFunc(it)
+    }
+    new StateDStream(self, newUpdateFunc, partitioner,
+      rememberPartitioner, Some(initialRDD))
```
Comment (zsxwing) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/PairDStreamFunctions.scala]: nit: can be just one line

File: streaming/src/test/scala/org/apache/spark/streaming/DStreamClosureSuite.scala [hunk 11]
```diff
@@ -171,6 +171,10 @@ class DStreamClosureSuite extends SparkFunSuite with BeforeAndAfterAll {
   private def testUpdateStateByKey(ds: DStream[(Int, Int)]): Unit = {
     val updateF1 = (_: Seq[Int], _: Option[Int]) => { return; Some(1) }
     val updateF2 = (_: Iterator[(Int, Seq[Int], Option[Int])]) => { return; Seq((1, 1)).toIterator }
+    val updateF3 = (_: Time, _: Int, _: Seq[Int], _: Option[Int]) => {
+      return
```
Comment (zsxwing) [inline streaming/src/test/scala/org/apache/spark/streaming/DStreamClosureSuite.scala]: return?

**Comment by zsxwing** [comment]
Just some style issues. Almost LGTM. I would like to let @tdas take a final look as this is adding new API.

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/PairDStreamFunctions.scala (hunk 6 above)
Comment (aramesh117) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/PairDStreamFunctions.scala]: Done.

**Comment by aramesh117** [comment]
I've fixed your comments. Sorry for the delay.

**Comment by aramesh117** [comment]
@tdas Have you had a chance to look at the pull request?

**Comment by aramesh117** [comment]
@zsxwing Seems like @tdas is busy. Can you merge this if possible before Spark 2.0?

**Comment by zzcclp** [comment]
@aramesh117 @zsxwing @tdas Is there any process at this pr?

**Comment by zzcclp** [comment]
ping @zsxwing @tdas

**Comment by aramesh117** [comment]
@zzcclp This pr was completed and approved by @zsxwing but hasn't been merged for some reason. I've tried to reach @tdas but he hasn't responded. What should we do? Now, there appear to be merge conflicts.

**Comment by zzcclp** [comment]
@aramesh117 you should update code based on master branch to resovlve conflicts, and then, jenkins will  test this pr automatically.

**Comment by zsxwing** [comment]
@aramesh117 sorry that I forgot this PR. Could you update to resolve conflicts, please? Thanks!

**Comment by zzcclp** [comment]
ping @aramesh117

**Comment by aramesh117** [comment]
@zzcclp and @zsxwing I will fix conflicts now.

**Comment by aramesh117** [comment]
@zsxwing and @zzcclp please take a look when you can - I've fixed all conflicts. I've run all tests and verified that all tests relevant to my change have passed. I would greatly appreciate it if we could merge it in soon. Thank you very much.

**Comment by zzcclp** [comment]
It sounds good. @zsxwing could you take a look at this pr? does this pr plan to be merged into branch-2.1?

**Comment by zsxwing** [comment]
retest this please.

LGTM. Let's run the test again. Will merge after test passes.

**Comment by zsxwing** [comment]
Merging to master and 2.1. @aramesh117 apologize for the delay again. Thanks a lot for your patience.

**Comment by aramesh117** [comment]
@zsxwing and @zzcclp thank you so much. This is much appreciated. :)

**Pull request closed without merging**